flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/FanOutShardSubscriber.java

```diff
@@ -235,8 +243,13 @@ private boolean consumeAllRecordsFromKinesisShard(
 		String continuationSequenceNumber;
 
 		do {
-			// Read timeout will occur after 30 seconds, add a sanity timeout here to prevent lockup
-			FanOutSubscriptionEvent subscriptionEvent = queue.poll(DEQUEUE_WAIT_SECONDS, SECONDS);
+			FanOutSubscriptionEvent subscriptionEvent;
+			if (queue.isEmpty() && subscriptionErrorEvent.get() != null) {
```

Comment:
Question:
Can you briefly explain why we would wait until the queue is fully consumed before handling exceptions?

Reply:
Events are processed in chronological order, the consumer will process events in the order they arrive (since an error cancels the subscription, and no more records will be received on the same subscription). This approach is the same as using a queue with sufficient space, but avoids the scenario when the queue is full. For example, consider the following sequence:
 1. batch A 
 2. batch B
 3. batch C
 4. error

The 4 events will be processed in that order. In the previous approach we would have to discard of a batch when the queue is at capacity.

If the errors were processed immediately, that would likely mean cancelling the EFO subscription and reacquiring, which would likely result in us discarding the queued record batches.